x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/adapitiveallocations/AdaptiveAllocationsScaler.java

```diff
@@ -0,0 +1,140 @@
+
+    private static final double SCALE_UP_THRESHOLD = 0.9;
+    private static final double SCALE_DOWN_THRESHOLD = 0.85;
+
+    private static final Logger logger = LogManager.getLogger(AdaptiveAllocationsScaler.class);
+
+    private final String deploymentId;
+    private final KalmanFilter requestRateEstimator;
+    private final KalmanFilter inferenceTimeEstimator;
+
+    private int numberOfAllocations;
+    private Integer minNumberOfAllocations;
+    private Integer maxNumberOfAllocations;
+    private boolean dynamicsChanged;
+
+    AdaptiveAllocationsScaler(String deploymentId, int numberOfAllocations) {
+        this.deploymentId = deploymentId;
+        // A smoothing factor of 100 roughly means the last 100 measurements have an effect
+        // on the estimated values. The sampling time is 10 seconds, so approximately the
+        // last 15 minutes are taken into account.
+        // For the request rate, use auto-detection for dynamics changes, because the request
+        // rate maybe change due to changed user behaviour.
+        // For the inference time, don't use this auto-detection. The dynamics may change when
+        // the number of allocations changes, which is passed explicitly to the estimator.
+        requestRateEstimator = new KalmanFilter(deploymentId + ":rate", 100, true);
+        inferenceTimeEstimator = new KalmanFilter(deploymentId + ":time", 100, false);
+        this.numberOfAllocations = numberOfAllocations;
+        this.minNumberOfAllocations = null;
+        this.maxNumberOfAllocations = null;
+        this.dynamicsChanged = false;
+    }
+
+    void setMinMaxNumberOfAllocations(Integer minNumberOfAllocations, Integer maxNumberOfAllocations) {
+        this.minNumberOfAllocations = minNumberOfAllocations;
+        this.maxNumberOfAllocations = maxNumberOfAllocations;
+    }
+
+    void process(AdaptiveAllocationsScalerService.Stats stats, double timeIntervalSeconds, int numberOfAllocations) {
+        // The request rate (per second) is the request count divided by the time.
+        // Assuming a Poisson process for the requests, the variance in the request
+        // count equals the mean request count, and the variance in the request rate
+        // equals that variance divided by the time interval squared.
+        // The minimum request count is set to 1, because lower request counts can't
+        // be reliably measured.
+        // The estimated request rate should be used for the variance calculations,
+        // because the measured request rate gives biased estimates.
+        double requestRate = (double) stats.requestCount() / timeIntervalSeconds;
+        double requestRateEstimate = requestRateEstimator.hasValue() ? requestRateEstimator.estimate() : requestRate;
+        double requestRateVariance = Math.max(1.0, requestRateEstimate * timeIntervalSeconds) / Math.pow(timeIntervalSeconds, 2);
+        requestRateEstimator.add(requestRate, requestRateVariance, false);
```

Comment:
All inside the `if (stats.requestCount() > 0) {`?

Reply:
No, I don't think so. A measured request rate of 0 is a real thing.

(Otherwise if the request rate drops from high to 0 instantly, the filter is never updated, and no scale down is ever triggered)

Reply:
Are yes. I wasn't taking into account `requestRateEstimator` is actually updated here.